bin/monitor.ml

```diff
@@ -0,0 +1,131 @@
+open Import
+open Fiber.O
+module Client = Dune_rpc_client.Client
+
+let waiting_for_file_system_changes message =
+  Pp.seq message (Pp.verbatim ", waiting for filesystem changes...")
+
+let restarting_current_build message =
+  Pp.seq message (Pp.verbatim ", restarting current build...")
+
+let status progress job_number =
+  Console.Status_line.set
+    (Constant
+       (match (progress : Dune_rpc.Progress.t) with
+       | Waiting -> Pp.verbatim "Initializing..."
+       | In_progress { complete; remaining } ->
+         Pp.textf "Done: %u%% (%u/%u, %u left) (jobs: %u)"
+           (if complete + remaining = 0 then 0
+            else complete * 100 / (complete + remaining))
+           complete (complete + remaining) remaining job_number
+       | Interrupted ->
+         Pp.tag User_message.Style.Error (Pp.verbatim "Source files changed")
+         |> restarting_current_build
+       | Success ->
+         Pp.tag User_message.Style.Success (Pp.verbatim "Success")
+         |> waiting_for_file_system_changes
+       | Failed ->
+         Pp.tag User_message.Style.Error (Pp.verbatim "Had errors")
+         |> waiting_for_file_system_changes))
+
+let next_exn poller =
+  let+ next = Client.Stream.next poller in
+  match next with
+  | None -> User_error.raise [ Pp.text "RPC server disconnected." ]
+  | Some payload -> payload
+
+let poll_exn ?id client sub =
+  let+ poller = Client.poll ?id client sub in
+  match poller with
+  | Error e -> raise (Dune_rpc_private.Version_error.E e)
+  | Ok poller -> poller
+
+(** Shared state *)
+
+let job_number : int Fiber.Svar.t = Fiber.Svar.create 0
+
+let all_diagnostics : Dune_rpc.Diagnostic.Set.t Fiber.Svar.t =
+  Fiber.Svar.create Dune_rpc.Diagnostic.Set.empty
+
+(** Functions on state *)
+
+let display_diagnostics () =
+  Console.reset ();
+  (* TODO should this be parallel or sequential? *)
+  Fiber.parallel_iter_set
+    (module Dune_rpc.Diagnostic.Set)
+    (Fiber.Svar.read all_diagnostics)
+    ~f:(fun err ->
+      Fiber.return
+      @@ Console.print_user_message (Dune_rpc.Diagnostic.to_user_message err))
+
+let maybe_reset_diagnostics progress =
+  match (progress : Dune_rpc.Progress.t) with
+  | Interrupted ->
+    Console.reset ();
+    Fiber.Svar.write all_diagnostics Dune_rpc.Diagnostic.Set.empty
+  | _ -> Fiber.return ()
+
+(** Loops on poll updates *)
+
+let rec progress_loop (progress_poller : Dune_rpc.Progress.t Client.Stream.t) =
+  let* progress = next_exn progress_poller in
+  let job_number = Fiber.Svar.read job_number in
+  status progress job_number;
+  maybe_reset_diagnostics progress >>> progress_loop progress_poller
+
+let rec job_loop (jobs_poller : Dune_rpc.Job.Event.t list Client.Stream.t) =
+  let* jobs = next_exn jobs_poller in
+  Fiber.Svar.write job_number (List.length jobs) >>> job_loop jobs_poller
```

Comment:
`jobs` are start/stop events for jobs. So you need to keep an internal set of running jobs to know how many are running.

Reply:
Yes, I've noticed this. I was wondering why the running jobs count was differing so much.

Reply:
I've changed it to increment and decrement the job count now.